aws/internal/generators/listpages/main.go

```diff
@@ -0,0 +1,238 @@
+	err := header.Execute(&g.buf, headerInfo)
+	if err != nil {
+		log.Fatalf("error writing header: %s", err)
+	}
+}
+
+func (g *Generator) parsePackage(sourcePackage string) {
+	cfg := &packages.Config{
+		Mode: packages.NeedName | packages.NeedSyntax,
+	}
+	pkgs, err := packages.Load(cfg, sourcePackage)
+	if err != nil {
+		log.Fatal(err)
+	}
+	if len(pkgs) != 1 {
+		log.Fatalf("error: %d packages found", len(pkgs))
+	}
+	g.addPackage(pkgs[0])
+}
+
+func (g *Generator) addPackage(pkg *packages.Package) {
+	g.pkg = &Package{
+		name:  pkg.Name,
+		files: make([]*PackageFile, len(pkg.Syntax)),
+	}
+
+	for i, file := range pkg.Syntax {
+		g.pkg.files[i] = &PackageFile{
+			file: file,
+		}
+	}
+}
+
+type FuncSpec struct {
+	Name       string
+	RecvType   string
+	ParamType  string
+	ResultType string
+}
+
+func (g *Generator) generateFunction(functionName string) {
+	var function *ast.FuncDecl
+
+	// TODO: check if a Pages() function has been defined
+	for _, file := range g.pkg.files {
+		if file.file != nil {
+			for _, decl := range file.file.Decls {
+				if funcDecl, ok := decl.(*ast.FuncDecl); ok {
+					if funcDecl.Name.Name == functionName {
+						function = funcDecl
+						break
+					}
+				}
+			}
+			if function != nil {
+				break
+			}
+		}
+	}
+
+	if function == nil {
+		log.Fatalf("function \"%s\" not found", functionName)
+	}
+
+	funcSpec := FuncSpec{
+		Name:       function.Name.Name,
+		RecvType:   g.expandTypeField(function.Recv),
+		ParamType:  g.expandTypeField(function.Type.Params),  // Assumes there is a single input parameter
+		ResultType: g.expandTypeField(function.Type.Results), // Assumes we can take the first return parameter
+	}
+
+	err := g.tmpl.Execute(&g.buf, funcSpec)
+	if err != nil {
+		log.Fatalf("error writing function \"%s\": %s", functionName, err)
+	}
+}
+
+func (g *Generator) expandTypeField(field *ast.FieldList) string {
+	typeValue := field.List[0].Type
+	if star, ok := typeValue.(*ast.StarExpr); ok {
+		return fmt.Sprintf("*%s", g.expandTypeExpr(star.X))
+	}
+
+	log.Fatalf("Unexpected type expression: (%[1]T) %[1]v", typeValue)
+	return ""
+}
+
+func (g *Generator) expandTypeExpr(expr ast.Expr) string {
+	if ident, ok := expr.(*ast.Ident); ok {
+		return fmt.Sprintf("%s.%s", g.pkg.name, ident.Name)
+	}
+
+	log.Fatalf("Unexpected expression: (%[1]T) %[1]v", expr)
+	return ""
+}
+
+const headerTemplate = `// Code generated by "aws/internal/generators/listpages/main.go {{ .Parameters }}"; DO NOT EDIT.
+
+package {{ .DestinationPackage }}
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"{{ .SourcePackage }}"
+)
+`
+
+const functionTemplate = `
+
+func {{ .Name }}Pages(conn {{ .RecvType }}, input {{ .ParamType }}, fn func({{ .ResultType }}, bool) bool) error {
+	for {
+		output, err := conn.{{ .Name }}(input)
+		if err != nil {
+			return err
+		}
+
+		lastPage := aws.StringValue(output.NextToken) == ""
+		if !fn(output, lastPage) || lastPage {
+			break
+		}
+
+		input.NextToken = output.NextToken
```

Comment:
This field isn't always named `NextToken`, for example [RDS Cluster Parameters](https://docs.aws.amazon.com/sdk-for-go/api/service/rds/#DescribeDBClusterParametersOutput) use `Marker'.

Reply:
Ah, good to know. Thanks!

We'll probably need some kind of customization spec. Or punt and do it manually :)